### PR TITLE
Add SLO for PostgreSQL by VSHN

### PR DESCRIPTION
**Pull request by glrf**
This PR adds the uptime SLO for VSHNPostgreSQL as well as base boilerplate to generate.

In detail this PR:

* Adds [Sloth](https://sloth.dev) as a dependency that is run as a compile step
* Uses metrics exposed by [the SLI Exporter](https://github.com/vshn/appcat-sli-exporter)
* Uses the added sloth to generate recording rules and alerts for PostgreSQL by VSHN based on Google's SLO implementation and [multi window multi burn alerts framework](https://landing.google.com/sre/workbook/chapters/alerting-on-slos/#6-multiwindow-multi-burn-rate-alerts).
* Adds a first iteration of a runbook that should help an engineer if he receives the alert.

## Checklist

- [x] The PR has a meaningful title. It will be used to auto generate the
      changelog.
      The PR has a meaningful description that sums up the change. It will be
      linked in the changelog.
- [x] PR contains a single logical change (to build a better changelog).
- [x] Update the documentation. *Don't forget to generate the CRD API!*
- [x] Categorize the PR by adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog.

### CODE REVIEW

File: class/defaults.yml [hunk 1]
```diff
@@ -78,6 +82,26 @@ parameters:
...
+      vshn:
+        postgres:
+          uptime:
+            objective: 99.9
+            alerting:
+              page_alert:
+                # This should reduce non actionable alerts because of single instance restarts.
+                # The page alert looks (ammong other things) at the burn rate over the last 5min.
+                # If the alert is pending for more than 5m this indicates a real problem.
+                for: 6m
```
Comment (zugao) [inline class/defaults.yml:102]: We said a restart usually is about 1 min, any reasons we have 6m here?

File: component/slos.libsonnet [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+{
+  slothInput: {
+    'vshn-postgresql': [
+      newSLO('uptime', 'vshn-postgresql', params.slos.vshn.postgres.uptime) {
+        description: 'Uptime SLO for PostgreSQL by VSHN',
+        sli: {
+          events: {
+            // The  0*rate(...) makes sure that the query reports an error rate for all instances, even if that instance has never produced a single error
+            error_query: 'sum(rate(appcat_probes_seconds_count{reason!="success", service="VSHNPostgreSQL"}[{{.window}}]) or 0*rate(appcat_probes_seconds_count{service="VSHNPostgreSQL"}[{{.window}}]))  by (service, namespace, name)',
```
Comment (zugao) [inline component/slos.libsonnet:67]: If this fails for a database instance, it will fire for how long? Until the end of the month?

File: docs/modules/ROOT/pages/references/slo-parameters.adoc [hunk 3]
```diff
@@ -29,3 +29,71 @@ type:: dict
...
+
+Labels that will be added to any critical SLO alerts.
+`page_alerts` are alerts are critical alerts for a high burn rate that require immediate attention.
+
+== `alerting.tickt_labels`
+[horizontal]
+type:: dict
+
+Labels that will be added to any non-crtiical SLO alerts.
+`ticket_alerts` are alerts are alerts for an elevated burn rate that might require attention, but aren’t urg
```
Comment (zugao) [inline docs/modules/ROOT/pages/references/slo-parameters.adoc:51]:
```suggestion
`ticket_alerts` are alerts for an elevated burn rate that might require attention, but aren’t urgent.
```

File: docs/modules/ROOT/pages/references/slo-parameters.adoc [hunk 4]
```diff
@@ -29,3 +29,71 @@ type:: dict
...
+default:: 99.9
+
+The SLO objective in percentage between 0-100.
+
+=== `vshn.postgres.uptime.alerting.page_alert`
+[horizontal]
+type:: dict
+
+The alert configuration for `page_alerts`.
+Can be used add additional lables or annotations throght the `labels` or `annotations` key, disable the alert through the `disable` key, or delay emitting the alert throug the `for` key.
```
Comment (zugao) [inline docs/modules/ROOT/pages/references/slo-parameters.adoc:69]:
```suggestion
Can be used to add additional labels or annotations through `labels` or `annotations` key, disable the alert through the `disable` key, or delay emitting the alert through the `for` key.
```
This is a bit confusing, I corrected the typos but still not very clear. Can you rephrase that?

File: docs/modules/ROOT/pages/references/slo-parameters.adoc [hunk 5]
```diff
@@ -29,3 +29,71 @@ type:: dict
...
+
+=== `vshn.postgres.uptime.alerting.ticket_alert`
+[horizontal]
+type:: dict
+
+The alert configuration for `page_alerts`.
```
Comment (zugao) [inline docs/modules/ROOT/pages/references/slo-parameters.adoc:75]:
```suggestion
The alert configuration for `ticket_alert`.
```

File: docs/modules/ROOT/pages/references/slo-parameters.adoc [hunk 6]
```diff
@@ -29,3 +29,71 @@ type:: dict
...
+Can be used add additional lables or annotations throght the `labels` or `annotations` key, disable the alert through the `disable` key, or delay emitting the alert throug the `for` key.
```
Comment (zugao) [inline docs/modules/ROOT/pages/references/slo-parameters.adoc:76]: Again the same typos.

File: docs/modules/ROOT/pages/runbooks/vshn-postgresql.adoc [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+If you received this alert, please add any insights you gained to improve this runbook.
+
+=== icon:glasses[] Overview
+
+The SLI measures the uptime of each PostgreSQL by VSHN instance.
+This SLI is measured by a prober that executes a SQL query every second.
+
+If this SLI results in an alert, it means that a significant number of SQL queries failed and that we risk missing the SLO.
+
+There are two types of alerts that fire if we expect to miss the configured objective.
```
Comment (zugao) [inline docs/modules/ROOT/pages/runbooks/vshn-postgresql.adoc:18]: why these are called page and ticket alert?

File: docs/modules/ROOT/pages/runbooks/vshn-postgresql.adoc [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+Further it can help to look at the `Object` resources that are created for this instance.
+
+[source,shell]
+----
+kubectl --as cluster-admin get object -l crossplane.io/composite=$COMPOSITE
+----
+
+If any of them are not synced, describing them should point you in the right direction.
+
+Finally, it might also be helpful to look at the logs of various crossplane components in namespace `sny-crossplane`.
```
Comment (zugao) [inline docs/modules/ROOT/pages/runbooks/vshn-postgresql.adoc:69]:
```suggestion
Finally, it might also be helpful to look at the logs of various crossplane components in namespace `syn-crossplane`.
```

File: docs/modules/ROOT/pages/runbooks/vshn-postgresql.adoc [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+----
+export INSTANCE_NAMESPACE=$(kubectl -n $NAMESPACE get vshnpostgresql $NAME -o jsonpath="{.status.instanceNamespace}")
+kubectl --as cluster-admin -n $INSTANCE_NAMESPACE get pod
+----
+
+If they're running, check the logs if there are any obvious error messages
+
+[source,shell]
+----
+kubectl --as cluster-admin -n $INSTANCE_NAMESPACE logs ${COMPOSITE}-0
```
Comment (zugao) [inline docs/modules/ROOT/pages/runbooks/vshn-postgresql.adoc:87]: Nitpick: you might want to grab the pods by labels instead of a specific pod.

File: run-sloth [hunk 10]
```diff
@@ -0,0 +1,65 @@
+#!/bin/bash
```
Comment (zugao) [inline run-sloth:1]: Can you explain in docs what is this script about? what's the purpose and why it exists?

**Review by zugao — COMMENTED** [review]
The runbook is great so far, there are some typos and maybe it would be great to have a doc for the script.

File: class/defaults.yml (hunk 1 above)
Comment (glrf) [reply class/defaults.yml:102]: Because the alert uses the rate over 5 minutes. So a ~30s might trigger the alert for 5min. Not completely happy with this solution, but I don't have a better idea right now

Comment (zugao) [reply class/defaults.yml:102]: we cannot diminish the rate to say 2 min? It's not feasible from your point of view?

File: component/slos.libsonnet (hunk 2 above)
Comment (glrf) [reply component/slos.libsonnet:67]: No. For a better explanation see: https://sre.google/workbook/alerting-on-slos/#6-multiwindow-multi-burn-rate-alerts

TL;DR: critical alerts should disappear after at most 30min after the errors stopped. Ticket alerts might stick around for 6h after everything is resolved

File: docs/modules/ROOT/pages/runbooks/vshn-postgresql.adoc (hunk 7 above)
Comment (glrf) [reply docs/modules/ROOT/pages/runbooks/vshn-postgresql.adoc:18]: Based on https://sre.google/workbook/alerting-on-slos/#6-multiwindow-multi-burn-rate-alerts and Sloths terminology. I.e. "Ticket" alerts should in general result in a ticket or some notification in a dashboard and not directly page someone.

File: class/defaults.yml (hunk 1 above)
Comment (glrf) [reply class/defaults.yml:102]: No that's defined by Sloth and just how Google's SLO alerting framework should work.

File: run-sloth (hunk 10 above)
Comment (glrf) [reply run-sloth:1]: Is this  clear enough for you?

Comment (zugao) [reply run-sloth:1]: Now it's better. I know precisely where to look for before analysing the code.

**Review by zugao — APPROVED** [review]
LGTM
